File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    int check(vector<int>mp)
    {
        int count=0;
        for(int i=0;i<mp.size();i++)
        {
            if(mp[i]>1)
            {
                return -1;
            }
            else if(mp[i]==1)
                count++;
        }
        return count;
    }
    void recur(vector<string>&arr,vector<int>mp,int i,int &ans)
    {
        
        if(check(mp)==-1)
        {
            return;
        }
        else{
            ans=max(ans,check(mp));
        }
        if(i>=arr.size())
            return ;
        recur(arr,mp,i+1,ans);
        for(int j=0;j<arr[i].size();j++)
        {
            mp[arr[i][j]-'a']++;
        }
        recur(arr,mp,i+1,ans);
    }
    int maxLength(vector<string>& arr) {
        int ans=0;
        vector<int>mp(26,0);
        recur(arr,mp,0,ans);
        return ans;
    }
};

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.java
class Solution {
    public int recur(List<String>arr,int curr,int val,int []mp){
        int ans=val;
        
        
        for(int i=curr;i<arr.size();i++){
            // System.out.println(curr+" "+i);
            int []currMp=new int[26];
            for(int cnt=0;cnt<26;cnt++){
                currMp[cnt]=mp[cnt];
            }
            int flag=0;
            for(int j=0;j<arr.get(i).length();j++){
                char ch=arr.get(i).charAt(j);
                if(currMp[ch-'a']==1){
                    flag=1;
                    break;
                }else{
                    currMp[ch-'a']++;
                }
            }
            if(flag==0){
                ans=Math.max(ans,recur(arr,i+1,val+arr.get(i).length(),currMp));
            }
        }
        return ans;
        
    }
    public int maxLength(List<String> arr) {
        int [] mp=new int[26];
        return recur(arr,0,0,mp);
          
    }
}